1.segment_images/scripts/0.segment_nuclei_organoids.py

```diff
@@ -0,0 +1,191 @@
+
+
+# In[5]:
+
+
+# make a 2.5 D max projection image stack with a sliding window of 3 slices
+
+image_stack_2_5D = np.empty((0, imgs.shape[1], imgs.shape[2]), dtype=imgs.dtype)
+for image_index in range(imgs.shape[0]):
+    image_stack_window = imgs[image_index : image_index + window_size]
+    if not image_stack_window.shape[0] == window_size:
+        break
+    # max project the image stack
+    image_stack_2_5D = np.append(
+        image_stack_2_5D, np.max(image_stack_window, axis=0)[np.newaxis, :, :], axis=0
+    )
+
+image_stack_2_5D = np.array(image_stack_2_5D)
+imgs = np.array(image_stack_2_5D)
+print("2.5D image stack shape:", image_stack_2_5D.shape)
+
+
+# ## Cellpose
+
+# In[6]:
+
+
+# Load the model
+model_name = "nuclei"
+use_GPU = core.use_gpu()
+model = models.CellposeModel(gpu=use_GPU, model_type=model_name)
+
+# Perform segmentation
+# labels, details, _ = model.eval(imgs, diameter=50, channels=[0, 0])
+labels, details, _ = model.eval(
+    imgs, diameter=75, channels=[0, 0], z_axis=0, stitch_threshold=0.8
+)
+
+
+# <img src="../notebook_imgs/Sliding_window_unaggregate.jpg" alt="image" width="300"/>
+
+# ## Reverse the sliding window max projection
+
+# In[7]:
+
+
+# reverse sliding window max projection
+full_mask_z_stack = []
+reconstruction_dict = {index: [] for index in range(original_z_slice_count)}
+print(f"Decoupling the sliding window max projection of {window_size} slices")
+
+for z_stack_mask_index in range(len(labels)):
+    z_stack_decouple = []
+    [z_stack_decouple.append(labels[z_stack_mask_index]) for _ in range(window_size)]
+    for z_window_index, z_stack_mask in enumerate(z_stack_decouple):
+        if not (z_stack_mask_index + z_window_index) >= original_z_slice_count:
+            reconstruction_dict[z_stack_mask_index + z_window_index].append(
+                z_stack_mask
+            )
+
+
+# In[8]:
+
+
+# max project each z position back to the original image
+reconstructed_masks = np.empty(
+    (0, labels[0].shape[0], labels[0].shape[1]), dtype=labels[0].dtype
+)
+
+for z_index in range(original_z_slice_count):
+    z_stack_masks = np.array(reconstruction_dict[z_index])
+    z_stack_max_projected = np.max(z_stack_masks, axis=0)[np.newaxis, :, :]
+    reconstructed_masks = np.append(reconstructed_masks, z_stack_max_projected, axis=0)
+
+print(reconstructed_masks.shape)
+if in_notebook:
+    # show each z slice of the image and masks
+    for z in range(reconstructed_masks.shape[0]):
+        fig = plt.figure(figsize=(10, 5))
+        plt.subplot(121)
+        plt.imshow(original_imgs[z, :, :], cmap="gray")
+        plt.title("image")
+        plt.axis("off")
+        plt.subplot(122)
+        plt.imshow(reconstructed_masks[z], cmap="magma")
+        plt.title("masks")
+        plt.axis("off")
+        plt.show()
+
+
+# In[9]:
+
+
+# # save the masks
+print(reconstructed_masks.shape)
+# save the masks as tiff
+mask_file_path = pathlib.Path(mask_path / "nuclei_masks.tiff").resolve()
```

Comment:
Is this saving the masks per image set? This looks hardcoded to save the masks as one file name. This code might just be practice code for one image set, but recommend already setting up the code to save the masks as unique names now. You can certainly say that is out-of-scope, your choice! :)

Reply:
The file name is hardcoded, but the path is not. The path is dynamic based on the path specified in the args